tests/integration/test_charm.py

```diff
@@ -201,3 +202,95 @@ def test_job_submission_works(juju: jubilant.Juju) -> None:
     # Get the hostname of the compute node from a Slurm job.
     sackd_result = juju.exec(f"srun --partition {SLURMD_APP_NAME} hostname -s", unit=sackd_unit)
     assert sackd_result.stdout == slurmd_result.stdout
+
+
+@pytest.mark.order(10)
+def test_gpu_job_submission_works(juju: jubilant.Juju) -> None:
+    """Test that a job requesting a GPU can be successfully submitted to the Slurm cluster."""
+    sackd_unit = f"{SACKD_APP_NAME}/0"
+    slurmd_unit = f"{SLURMD_APP_NAME}/0"
+
+    # Set up a mock GPU device on the slurmd unit by mounting over relevant files in /sys and /proc
+    # WARNING: This is tightly coupled to the method the Slurm "Autodetect=nvidia" plugin uses to
+    # detect GPUs. Changes to that method may break this test.
+    # Mock NUMA region info in /sys
+    juju.exec("mkdir -p /tmp/sys/bus/pci/drivers/nvidia/0000:01:00.0/", unit=slurmd_unit)
+    juju.exec(
+        "cp /sys/devices/system/node/node0/cpulist /tmp/sys/bus/pci/drivers/nvidia/0000:01:00.0/local_cpulist",
+        unit=slurmd_unit,
+    )
+    juju.exec(
+        "sudo mount -t overlay overlay -o lowerdir=/sys/bus/pci/drivers:/tmp/sys/bus/pci/drivers /sys/bus/pci/drivers",
+        unit=slurmd_unit,
+    )
+
+    # Mock GPU info in /proc
+    gpu_information = textwrap.dedent(
+        """\
+    Model: 		 Mock GPU
+    IRQ:   		 185
+    GPU UUID: 	 GPU-12345678-90ab-cdef-1234-567890abcdef
+    Video BIOS: 	 12.34.56.78.aa
+    Bus Type: 	 PCIe
+    DMA Size: 	 47 bits
+    DMA Mask: 	 0x7fffffffffff
+    Bus Location: 	 0000:01:00.0
+    Device Minor: 	 0
+    GPU Firmware: 	 123.456.78
+    GPU Excluded:	 No
+    """
+    )
+    juju.exec("mkdir -p /tmp/proc/driver/nvidia/gpus/0000:01:00.0/", unit=slurmd_unit)
+    juju.exec(
+        f"echo '{gpu_information}' > /tmp/proc/driver/nvidia/gpus/0000:01:00.0/information",
+        unit=slurmd_unit,
+    )
+    # Can't Overlay mount with /proc. Attempts fail with error:
+    #   "wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error"
+    # Bind mount over the top instead. This should only block `/proc/driver/rtc` briefly
+    juju.exec("sudo mount --bind /tmp/proc/driver /proc/driver", unit=slurmd_unit)
+
+    # Slurm needs a GPU device file - use /dev/zero as a mock
+    juju.exec("sudo touch /dev/nvidia0", unit=slurmd_unit)  # mount point must exist
```

Comment:
This in-line comment feels lost. Let's update the comment above instead.

```suggestion
    # Slurm expects a GPU device file to be under /dev when auto-detecting GPUs.
    # Use /dev/zero as a mock by bind mounting over an existing /dev/nvidia0 device file.
    juju.exec("sudo touch /dev/nvidia0", unit=slurmd_unit)
```

Feel free to wordsmith as necessary but let's not use in-line comments.

Reply:
I've reworded this comment with some minor wordsmithing. I do think we should be open to using in-line comments where appropriate, however. They're useful for literal side remarks like "oh, by the way, don't forget you need an empty file to exist as the mount point before the bind will work".

Reply:
Agreed they shouldn't describe a block of statements but they're useful for making it easier to grasp a single line, like a quick reminder "`touch /dev/nvidia0` is necessary, don't refactor this line out". Still, I'm happy with the more descriptive comment we've got now - I'm just objecting to "let's not use in-line comments" as a hard and fast rule for the codebase, but sounds like that's not what you meant.